sweet/benchmarks/cockroachdb/main.go

```diff
@@ -0,0 +1,611 @@
+	)
+	inst.cmd.Stdout = &inst.output
+	inst.cmd.Stderr = &inst.output
+	if err := inst.cmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start instance %q: %v", inst.name, err)
+	}
+	return instances, nil
+}
+
+func launchCockroachCluster(cfg *config) ([]*cockroachdbInstance, error) {
+	if cfg.bench.nodeCount == 1 {
+		// Use `cockroach start-single-node` instead for single node clusters.
+		return launchSingleNodeCluster(cfg)
+	}
+	var instances []*cockroachdbInstance
+	for i := 0; i < cfg.bench.nodeCount; i++ {
+		instances = append(instances, &cockroachdbInstance{
+			name:     fmt.Sprintf("roach-node-%d", i+1),
+			sqlPort:  basePort + 2*i,
+			httpPort: basePort + 2*i + 1,
+		})
+	}
+
+	// Start the instances with `cockroach start`.
+	for n, inst := range instances {
+		allOtherInstances := append([]*cockroachdbInstance{}, instances[:n]...)
+		allOtherInstances = append(allOtherInstances, instances[n+1:]...)
+		join := fmt.Sprintf("--join=%s", clusterAddresses(allOtherInstances))
+
+		inst.cmd = exec.Command(cfg.cockroachdbBin,
+			"start",
+			"--insecure",
+			"--listen-addr", inst.sqlAddr(),
+			"--http-addr", inst.httpAddr(),
+			"--cache", cacheSize,
+			"--store", fmt.Sprintf("%s/%s", cfg.tmpDir, inst.name),
+			"--logtostderr",
+			join,
+		)
+		inst.cmd.Env = append(os.Environ(),
+			fmt.Sprintf("GOMAXPROCS=%d", cfg.procsPerInst),
+		)
+		inst.cmd.Stdout = &inst.output
+		inst.cmd.Stderr = &inst.output
+		if err := inst.cmd.Start(); err != nil {
+			return nil, fmt.Errorf("failed to start instance %q: %v", inst.name, err)
+		}
+	}
+
+	// Initialize the cluster with `cockroach init`.
+	inst1 := instances[0]
+	initCmd := exec.Command(cfg.cockroachdbBin,
+		"init",
+		"--insecure",
+		fmt.Sprintf("--host=%s", cfg.host),
+		fmt.Sprintf("--port=%d", inst1.sqlPort),
+	)
+	initCmd.Env = append(os.Environ(),
+		fmt.Sprintf("GOMAXPROCS=%d", cfg.procsPerInst),
+	)
+	initCmd.Stdout = &inst1.output
+	initCmd.Stderr = &inst1.output
+	if err := initCmd.Start(); err != nil {
+		return nil, fmt.Errorf("failed to init instance %q: %v", inst1.name, err)
+	}
+	return instances, nil
+}
+
+// waitForCluster pings nodes in the cluster until one responds, or
+// we time out. We only care to wait for one node to respond as the
+// workload will work as long as it can connect to one node initially.
+// The --ramp flag will take care of startup noise.
+func waitForCluster(instances []*cockroachdbInstance, cfg *config) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	for _, inst := range instances {
+		inst := inst
+		go func(ctx context.Context) {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					// The node will almost certainly not be ready right away, so wait
+					// 5 seconds first and between pings. 5 seconds was chosen through
+					// trial and error as a time that nodes are *usually* ready by.
+					time.Sleep(5 * time.Second)
```

Comment:
nit: instead of the `time.Sleep` in a `default` case, a `case <-time.After(5 * time.Second)` would allow this loop to react more quickly to context cancellation.

Reply:
Done! Thanks for the tip 😄